File: js/src/ssmpl.js
#!/usr/bin/env node
// SSMPL — Static Site · Multi-Password Lock  (JavaScript / WebCrypto)
// Works in Node >= 18 (globalThis.crypto is WebCrypto) and in browsers.
//
// Backend (encrypt / 密本 / multi-password) and Frontend (unlock / decrypt) share this one module.
// The online payload is the blob JSON; a reader needs a valid password to recover the plaintext.
//
// Config (settable via setConfig / config object):
//   hash       : "SHA-256" | "SHA-512"
//   iterations : int (KDF cost)
//   cipher     : "AES-256-GCM" | "AES-256-CBC-HMAC-SHA256"
//   salt       : Uint8Array (public salt)

const subtle = (globalThis.crypto || {}).subtle;
if (!subtle) throw new Error('SSMPL: WebCrypto (crypto.subtle) is required.');

const DEFAULT_CONFIG = {
  hash: 'SHA-256',
  iterations: 250000,
  cipher: 'AES-256-GCM',
  salt: null, // must be set (16 bytes) before use; or a fixed salt provided by caller
};

// ---------- bytes / base64 ----------
function b64e(bytes) {
  let s = '';
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}
function b64d(b64) {
  const bin = atob(b64);
  const u = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) u[i] = bin.charCodeAt(i);
  return u;
}
function xorBytes(a, b) {
  const out = new Uint8Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] ^ b[i];
  return out;
}
function toBytes(s) { return new TextEncoder().encode(s); }
function fromBytes(u) { return new TextDecoder().decode(u); }
function normalizeConfig(cfg) {
  const c = Object.assign({}, DEFAULT_CONFIG, cfg || {});
  if (!c.salt || c.salt.length < 8) throw new Error('SSMPL: config.salt must be set (>=8 bytes).');
  return c;
}

// ---------- one-way KDF: P = PBKDF2(password, salt, iterations, 32) ----------
async function deriveKey(password, cfg) {
  const c = normalizeConfig(cfg);
  const km = await subtle.importKey('raw', toBytes(String(password)), 'PBKDF2', false, ['deriveBits']);
  const bits = await subtle.deriveBits(
    { name: 'PBKDF2', salt: c.salt, iterations: c.iterations, hash: c.hash },
    km, 256);
  return new Uint8Array(bits); // P (32 bytes)
}

// ---------- two-way symmetric cipher ----------
async function macKeyFromK(K) {
  // deterministic mac key for CBC+HMAC mode (cross-language compatible)
  const label = toBytes('SSMPL.HMAC');
  const data = new Uint8Array(K.length + label.length);
  data.set(K); data.set(label, K.length);
  const d = await subtle.digest('SHA-256', data);
  return new Uint8Array(d);
}
async function hmac(key, msg) {
  const k = await subtle.importKey('raw', key, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
  const sig = await subtle.sign('HMAC', k, msg);
  return new Uint8Array(sig);
}

async function encryptBytes(plain, K, cfg) {
  const c = normalizeConfig(cfg);
  if (c.cipher === 'AES-256-CBC-HMAC-SHA256') {
    const iv = crypto.getRandomValues(new Uint8Array(16));
    const encKey = await subtle.importKey('raw', K, { name: 'AES-CBC' }, false, ['encrypt']);
    const ct = new Uint8Array(await subtle.encrypt({ name: 'AES-CBC', iv }, encKey, toBytes(plain)));
    const mkey = await macKeyFromK(K);
    const macInput = new Uint8Array(iv.length + ct.length);
    macInput.set(iv); macInput.set(ct, iv.length);
    const mac = await hmac(mkey, macInput);
    return { iv: b64e(iv), ct: b64e(ct), mac: b64e(mac) };
  } else { // AES-256-GCM
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const key = await subtle.importKey('raw', K, { name: 'AES-GCM' }, false, ['encrypt']);
    const ct = new Uint8Array(await subtle.encrypt({ name: 'AES-GCM', iv }, key, toBytes(plain)));
    const tag = ct.slice(ct.length - 16);          // last 16 bytes = auth tag
    const body = ct.slice(0, ct.length - 16);
    return { iv: b64e(iv), ct: b64e(body), tag: b64e(tag) };
  }
}
async function decryptBytes(item, K, cfg) {
  const c = normalizeConfig(cfg);
  if (c.cipher === 'AES-256-CBC-HMAC-SHA256') {
    const mkey = await macKeyFromK(K);
    const macInput = new Uint8Array(b64d(item.iv).length + b64d(item.ct).length);
    macInput.set(b64d(item.iv)); macInput.set(b64d(item.ct), b64d(item.iv).length);
    const expect = await hmac(mkey, macInput);
    if (!timingEq(expect, b64d(item.mac))) throw new Error('SSMPL: MAC mismatch (wrong key).');
    const encKey = await subtle.importKey('raw', K, { name: 'AES-CBC' }, false, ['decrypt']);
    const plain = await subtle.decrypt({ name: 'AES-CBC', iv: b64d(item.iv) }, encKey, b64d(item.ct));
    return fromBytes(new Uint8Array(plain));
  } else { // AES-256-GCM
    const key = await subtle.importKey('raw', K, { name: 'AES-GCM' }, false, ['decrypt']);
    const ct = b64d(item.ct), tag = b64d(item.tag);
    const combined = new Uint8Array(ct.length + tag.length);
    combined.set(ct); combined.set(tag, ct.length);
    const plain = await subtle.decrypt({ name: 'AES-GCM', iv: b64d(item.iv) }, key, combined);
    return fromBytes(new Uint8Array(plain));
  }
}
function timingEq(a, b) {
  if (a.length !== b.length) return false;
  let d = 0;
  for (let i = 0; i < a.length; i++) d |= a[i] ^ b[i];
  return d === 0;
}

// ---------- 密本 = P XOR K ----------
function computeBook(P, K) { return xorBytes(P, K); }

// ---------- backend operations ----------
async function unlockKey(item, P, cfg) {
  // try each 密本: K = P XOR 密本, then verify by decrypting
  for (const bk of item.books) {
    const K = xorBytes(P, b64d(bk));
    try { await decryptBytes(item, K, cfg); return K; } catch (_) { /* next 密本 */ }
  }
  return null;
}

async function encryptItem(plaintext, password, cfg) {
  const c = normalizeConfig(cfg);
  const K = crypto.getRandomValues(new Uint8Array(32));  // random content key
  const item = await encryptBytes(plaintext, K, c);
  const P = await deriveKey(password, c);
  item.books = [b64e(computeBook(P, K))];
  return item;
}

async function getKey(item, password, cfg) {
  const c = normalizeConfig(cfg);
  const P = await deriveKey(password, c);
  return await unlockKey(item, P, c);
}

async function decryptItem(item, password, cfg) {
  const K = await getKey(item, password, cfg);
  if (!K) throw new Error('SSMPL: wrong password (no 密本 decrypts).');
  return decryptBytes(item, K, cfg);
}

async function addPassword(item, newPassword, unlockPassword, cfg) {
  const K = await getKey(item, unlockPassword, cfg);
  if (!K) throw new Error('SSMPL: unlockPassword is wrong.');
  const P = await deriveKey(newPassword, cfg);
  const next = JSON.parse(JSON.stringify(item)); // deep copy so we don't mutate caller's item
  if (!next.books) next.books = [];
  next.books.push(b64e(computeBook(P, K)));
  return next;
}

// Revoke access for a given password on ONE item: drop that password's 密本 (no re-encrypt).
async function removePassword(item, password, cfg) {
  const c = normalizeConfig(cfg);
  const P = await deriveKey(password, c);
  const next = JSON.parse(JSON.stringify(item));
  const keep = [];
  for (const bk of next.books || []) {
    const K = xorBytes(P, b64d(bk));
    let mine = false;
    try { await decryptBytes(next, K, c); mine = true; } catch (_) { }
    if (!mine) keep.push(bk);                     // a book that decrypts = this password's -> drop
  }
  next.books = keep;
  return next;
}

// ---------- frontend lock (load a blob, unlock with a password, decrypt) ----------
class Lock {
  constructor(blob, cfg) {
    this.blob = blob;               // the hosted JSON payload (authoritative source of truth)
    this.config = Object.assign({}, cfg || {}); // only user overrides live here
    this._P = null;                 // cached derived key after a successful unlock
    this._map = {};                 // name -> item
    (function (self) {
      Object.keys(blob.items || {}).forEach(function (n) { self._map[n] = blob.items[n]; });
    })(this);
  }
  setConfig(cfg) { this.config = Object.assign({}, this.config, cfg || {}); this._P = null; return this; }
  setHash(h) { this.config.hash = h; this._P = null; return this; }
  setIterations(n) { this.config.iterations = n; this._P = null; return this; }
  setCipher(c) { this.config.cipher = c; this._P = null; return this; }

  async unlock(password) {
    // Build the effective config from the blob, letting only explicit user overrides win.
    const cfg = Object.assign({}, DEFAULT_CONFIG, {
      salt: this.blob.salt ? b64d(this.blob.salt) : this.config.salt,
      hash: this.blob.hash || DEFAULT_CONFIG.hash,
      iterations: this.blob.iterations || DEFAULT_CONFIG.iterations,
      cipher: this.blob.cipher || DEFAULT_CONFIG.cipher,
    }, this.config);
    const names = Object.keys(this._map);
    const P = await deriveKey(String(password), cfg);
    for (const n of names) {
      const K = await unlockKey(this._map[n], P, cfg);
      if (K) { this._P = P; this._cfg = cfg; return true; }
    }
    return false;
  }
  isUnlocked() { return this._P !== null; }
  async decrypt(name) {
    if (!this._P) throw new Error('SSMPL: not unlocked yet.');
    const item = this._map[name];
    if (!item) throw new Error('SSMPL: item not found: ' + name);
    const K = await unlockKey(item, this._P, this._cfg);
    if (!K) throw new Error('SSMPL: cannot decrypt item (wrong key).');
    return decryptBytes(item, K, this._cfg);
  }
}

export {
  DEFAULT_CONFIG, b64e, b64d, deriveKey, encryptBytes, decryptBytes, computeBook,
  encryptItem, getKey, decryptItem, addPassword, removePassword, Lock, xorBytes,
};

File: cpp/src/ssmpl.cpp
// SSMPL implementation (OpenSSL / libcrypto). Mirrors js/src/ssmpl.js and python/ssmpl/core.py.
#include "ssmpl.hpp"

#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/sha.h>
#include <openssl/rand.h>
#include <stdexcept>
#include <cstring>

namespace ssmpl {

static const char* B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

std::string b64e(const std::vector<uint8_t>& in) {
    std::string out;
    size_t i = 0, n = in.size();
    while (i + 2 < n) {
        uint32_t v = (uint32_t(in[i]) << 16) | (uint32_t(in[i+1]) << 8) | in[i+2];
        out += B64[(v >> 18) & 63]; out += B64[(v >> 12) & 63];
        out += B64[(v >> 6) & 63]; out += B64[v & 63]; i += 3;
    }
    if (i + 1 == n) { uint32_t v = uint32_t(in[i]) << 16; out += B64[(v>>18)&63]; out += B64[(v>>12)&63]; out += "=="; }
    else if (i + 2 == n) { uint32_t v = (uint32_t(in[i])<<16)|(uint32_t(in[i+1])<<8); out += B64[(v>>18)&63]; out += B64[(v>>12)&63]; out += B64[(v>>6)&63]; out += "="; }
    return out;
}
std::vector<uint8_t> b64d(const std::string& in) {
    int val[256]; for (int i = 0; i < 256; i++) val[i] = -1;
    for (int i = 0; i < 64; i++) val[(unsigned char)B64[i]] = i;
    std::vector<uint8_t> out; uint32_t buf = 0; int bits = 0;
    for (char c : in) { if (c == '=' || c == '\n' || c == '\r') continue; int v = val[(unsigned char)c]; if (v < 0) continue; buf = (buf << 6) | v; bits += 6; if (bits >= 8) { bits -= 8; out.push_back((buf >> bits) & 0xFF); } }
    return out;
}
std::vector<uint8_t> xor_bytes(const std::vector<uint8_t>& a, const std::vector<uint8_t>& b) {
    std::vector<uint8_t> o(a.size());
    for (size_t i = 0; i < a.size(); i++) o[i] = a[i] ^ b[i];
    return o;
}
static const EVP_MD* md_from_hash(const std::string& h) { return (h == "SHA-512") ? EVP_sha512() : EVP_sha256(); }
static std::vector<uint8_t> pbkdf2(const std::string& pw, const std::vector<uint8_t>& salt, unsigned long iters, const char* hash) {
    const EVP_MD* md = (std::string(hash) == "SHA-512") ? EVP_sha512() : EVP_sha256();
    std::vector<uint8_t> out(32);
    PKCS5_PBKDF2_HMAC(pw.c_str(), (int)pw.size(), salt.data(), (int)salt.size(), (int)iters, md, 32, out.data());
    return out;
}

std::vector<uint8_t> derive_key(const std::string& password, const Config& cfg) {
    return pbkdf2(password, cfg.salt, cfg.iterations, cfg.hash.c_str());
}
static std::vector<uint8_t> mac_key(const std::vector<uint8_t>& K) {
    SHA256_CTX c; SHA256_Init(&c);
    SHA256_Update(&c, K.data(), K.size());
    static const unsigned char label[] = "SSMPL.HMAC";
    SHA256_Update(&c, label, sizeof(label) - 1);
    std::vector<uint8_t> d(SHA256_DIGEST_LENGTH); SHA256_Final(d.data(), &c);
    return d;
}
static std::vector<uint8_t> hmac(const std::vector<uint8_t>& key, const std::vector<uint8_t>& msg) {
    unsigned int len = 0;
    std::vector<uint8_t> out(EVP_MAX_MD_SIZE);
    HMAC(EVP_sha256(), key.data(), (int)key.size(), msg.data(), msg.size(), out.data(), &len);
    out.resize(len); return out;
}

Item encrypt_bytes(const std::string& plaintext, const std::vector<uint8_t>& K, const Config& cfg) {
    Item it;
    const std::vector<uint8_t> pt(plaintext.begin(), plaintext.end());
    if (cfg.cipher == "AES-256-CBC-HMAC-SHA256") {
        std::vector<uint8_t> iv(16); RAND_bytes(iv.data(), (int)iv.size());
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, K.data(), iv.data());
        std::vector<uint8_t> ct(pt.size() + 16); int len = 0, total = 0;
        EVP_EncryptUpdate(ctx, ct.data(), &len, pt.data(), (int)pt.size()); total = len;
        EVP_EncryptFinal_ex(ctx, ct.data() + total, &len); total += len;
        ct.resize(total); EVP_CIPHER_CTX_free(ctx);
        std::vector<uint8_t> macIn(iv.begin(), iv.end()); macIn.insert(macIn.end(), ct.begin(), ct.end());
        it.iv = b64e(iv); it.ct = b64e(ct); it.mac = b64e(hmac(mac_key(K), macIn));
    } else { // AES-256-GCM
        std::vector<uint8_t> iv(12); RAND_bytes(iv.data(), (int)iv.size());
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr);
        EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, (int)iv.size(), nullptr);
        EVP_EncryptInit_ex(ctx, nullptr, nullptr, K.data(), iv.data());
        std::vector<uint8_t> ct(pt.size() + 16); int len = 0, total = 0;
        EVP_EncryptUpdate(ctx, ct.data(), &len, pt.data(), (int)pt.size()); total = len;
        EVP_EncryptFinal_ex(ctx, ct.data() + total, &len); total += len;
        ct.resize(total);
        std::vector<uint8_t> tag(16); EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, 16, tag.data());
        EVP_CIPHER_CTX_free(ctx);
        it.iv = b64e(iv); it.ct = b64e(ct); it.tag = b64e(tag);
    }
    return it;
}
std::string decrypt_bytes(const Item& item, const std::vector<uint8_t>& K, const Config& cfg) {
    std::vector<uint8_t> iv = b64d(item.iv);
    if (cfg.cipher == "AES-256-CBC-HMAC-SHA256") {
        std::vector<uint8_t> ct = b64d(item.ct);
        std::vector<uint8_t> macIn = iv; macIn.insert(macIn.end(), ct.begin(), ct.end());
        if (hmac(mac_key(K), macIn) != b64d(item.mac)) throw std::runtime_error("SSMPL: MAC mismatch (wrong key).");
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_DecryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, K.data(), iv.data());
        std::vector<uint8_t> pt(ct.size() + 16); int len = 0, total = 0;
        EVP_DecryptUpdate(ctx, pt.data(), &len, ct.data(), (int)ct.size()); total = len;
        if (EVP_DecryptFinal_ex(ctx, pt.data() + total, &len) <= 0) { EVP_CIPHER_CTX_free(ctx); throw std::runtime_error("SSMPL: bad padding."); }
        total += len; pt.resize(total); EVP_CIPHER_CTX_free(ctx);
        return std::string(pt.begin(), pt.end());
    } else { // AES-256-GCM
        std::vector<uint8_t> ct = b64d(item.ct), tag = b64d(item.tag);
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr);
        EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, (int)iv.size(), nullptr);
        EVP_DecryptInit_ex(ctx, nullptr, nullptr, K.data(), iv.data());
        EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, (int)tag.size(), tag.data());
        std::vector<uint8_t> pt(ct.size() + 16); int len = 0, total = 0;
        EVP_DecryptUpdate(ctx, pt.data(), &len, ct.data(), (int)ct.size()); total = len;
        if (EVP_DecryptFinal_ex(ctx, pt.data() + total, &len) <= 0) { EVP_CIPHER_CTX_free(ctx); throw std::runtime_error("SSMPL: wrong key / auth failed."); }
        total += len; pt.resize(total); EVP_CIPHER_CTX_free(ctx);
        return std::string(pt.begin(), pt.end());
    }
}

std::vector<uint8_t> compute_book(const std::vector<uint8_t>& P, const std::vector<uint8_t>& K) { return xor_bytes(P, K); }

static std::vector<uint8_t> unlock_key(const Item& item, const std::vector<uint8_t>& P, const Config& cfg) {
    for (const auto& bk : item.books) {
        std::vector<uint8_t> K = xor_bytes(P, b64d(bk));
        try { decrypt_bytes(item, K, cfg); return K; } catch (...) { /* next 密本 */ }
    }
    return {};
}
Item encrypt_item(const std::string& plaintext, const std::string& password, const Config& cfg) {
    std::vector<uint8_t> K(32); RAND_bytes(K.data(), (int)K.size());
    Item it = encrypt_bytes(plaintext, K, cfg);
    std::vector<uint8_t> P = derive_key(password, cfg);
    it.books.push_back(b64e(compute_book(P, K)));
    return it;
}
std::vector<uint8_t> get_key(const Item& item, const std::string& password, const Config& cfg) {
    return unlock_key(item, derive_key(password, cfg), cfg);
}
std::string decrypt_item(const Item& item, const std::string& password, const Config& cfg) {
    std::vector<uint8_t> K = get_key(item, password, cfg);
    if (K.empty()) throw std::runtime_error("SSMPL: wrong password (no 密本 decrypts).");
    return decrypt_bytes(item, K, cfg);
}
Item add_password(const Item& item, const std::string& new_password, const std::string& unlock_password, const Config& cfg) {
    std::vector<uint8_t> K = get_key(item, unlock_password, cfg);
    if (K.empty()) throw std::runtime_error("SSMPL: unlock_password is wrong.");
    Item it = item;
    std::vector<uint8_t> P = derive_key(new_password, cfg);
    it.books.push_back(b64e(compute_book(P, K)));
    return it;
}
Item remove_password(const Item& item, const std::string& password, const Config& cfg) {
    std::vector<uint8_t> P = derive_key(password, cfg);
    Item it = item; it.books.clear();
    for (const auto& bk : item.books) {
        std::vector<uint8_t> K = xor_bytes(P, b64d(bk));
        bool mine = false;
        try { decrypt_bytes(item, K, cfg); mine = true; } catch (...) {}
        if (!mine) it.books.push_back(bk);
    }
    return it;
}

} // namespace ssmpl
